runtime/memprof.c

```diff
@@ -206,1126 +1258,921 @@ void caml_memprof_update_suspended(bool s) {
+        }
+      } else {
+        /* profile has changed, possibly stopped */
+        config = thread_config(thread);
+        /* any entries we had added, in the previous profile, have
+         * been orphaned and are no longer in the thread's table - we
+         * don't need to do anything with them.
+         * TODO: correct treatment of any Placeholder() entries when
+         * orphaning entries: delete them? */
+        new_entries = 0;
+        if (Sampling(config)) {
+          /* start again (for new profile) from the next sub-allocation. */
+          trigger_ofs = alloc_ofs - (rand_geom(domain) - 1);
+        } else {
+          /* not sampling; don't sample any more in this comballoc */
+          trigger_ofs = 0;
         }
       }
-      if (Is_exception_result(res)) break;
-      if (!stopped) allocs_sampled++;
+      if (cancelled)
+        break;
     }
-  }
+  } while (alloc_idx);
 
-  CAMLassert(alloc_ofs == 0 || Is_exception_result(res));
-  CAMLassert(allocs_sampled <= nallocs);
+  CAMLassert(alloc_ofs == 0 || cancelled);
+  CAMLassert(new_entries <= nallocs);
+
+  if (!cancelled) {
+    /* No exceptions were raised. Re-allocate the blocks in the minor
+     * heap. We must not trigger a GC after this point. */
 
-  if (!Is_exception_result(res)) {
-    /* The callbacks did not raise. The allocation will take place.
-       We now restore the minor heap in the state needed by
-       [Alloc_small_aux]. */
     if (Caml_state->young_ptr - whsize < Caml_state->young_trigger) {
       CAML_EV_COUNTER(EV_C_FORCE_MINOR_MEMPROF, 1);
-      caml_gc_dispatch();
+      caml_poll_gc_work();
     }
-
-    /* Re-allocate the blocks in the minor heap. We should not call the
-       GC after this. */
     Caml_state->young_ptr -= whsize;
-
-    /* Make sure this block is not going to be sampled again. */
-    shift_sample(whsize);
   }
 
-  /* Since [local->entries] is local to the current thread, we know for
-     sure that the allocated entries are the [alloc_sampled] last entries of
-     [local->entries]. */
-
-  for (i = 0; i < allocs_sampled; i++) {
-    uintnat idx = local->entries.len-allocs_sampled+i;
-    if (local->entries.t[idx].deleted) continue;
-    if (realloc_entries(&entries_global, 1)) {
-      /* Transfer the entry to the global array. */
-      struct tracked* t = &entries_global.t[entries_global.len];
-      entries_global.len++;
-      *t = local->entries.t[idx];
-
-      if (Is_exception_result(res)) {
-        /* The allocations are cancelled because of the exception,
-           but this callback has already been called. We simulate a
-           deallocation. */
-        t->block = Val_unit;
-        t->deallocated = 1;
+  /* Since [thread->entries] is local to the current thread, we know
+   * that the allocated entries are the [new_entries] last entries of
+   * [thread->entries]. Transfer them all to the per-domain entries
+   * table. */
+
+  for (size_t i = 0; i < new_entries; i++) {
```

Comment:
This loop is also complicated. Could it be expressed as one loop to update all the placeholders, then a call to `entries_transfer`?

Reply:
Only if we have also done your other suggestion, of running the callbacks from the previous loop by calling `entries_run_callbacks_exn` - because otherwise this might also transfer deferred allocation entries to the domain's entries table, where they may be called by other threads, counter to the documentation that allocation callbacks are called by the allocating thread if possible. So I'll look into this alongside that idea.